The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12271713341/job/34239075189) [(plain)](https://github.com/rust-lang/rust/commit/f8ca2e570fb1a6db18bef29475c3e017510bde97/checks/34239075189/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.1s done
#16 DONE 39.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/instrument-coverage/testprog.rs#LINUX stdout ----

error in revision `LINUX`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/instrument-coverage/testprog.LINUX/testprog.ll" "/checkout/tests/codegen/instrument-coverage/testprog.rs" "--check-prefix=CHECK" "--check-prefix" "LINUX" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100" "-DINSTR_PROF_DATA=__llvm_prf_data" "-DINSTR_PROF_NAME=__llvm_prf_names" "-DINSTR_PROF_CNTS=__llvm_prf_cnts" "-DINSTR_PROF_COVMAP=__llvm_covmap" "-DINSTR_PROF_COVFUN=__llvm_covfun" "-DCOMDAT_IF_SUPPORTED=, comdat"
--- stderr -------------------------------
/checkout/tests/codegen/instrument-coverage/testprog.rs:79:11: error: CHECK: expected string not found in input
/checkout/tests/codegen/instrument-coverage/testprog.rs:79:11: error: CHECK: expected string not found in input
// CHECK: @__covrec_{{[A-F0-9]+}}u = linkonce_odr hidden constant
          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/instrument-coverage/testprog.LINUX/testprog.ll:77:343: note: scanning from here
@__llvm_coverage_mapping = private constant { { i32, i32, i32, i32 }, [69 x i8] } { { i32, i32, i32, i32 } { i32 0, i32 69, i32 0, i32 6 }, [69 x i8] c"\02FBx\DA\E3\D5O\CEHM\CE\CE/-\D1\CFO\CA2G\F0JR\8BK\8A\F5\93\F3SR\D3S\F3\F43\F3\8AK\8AJsS\F3Jt\93\F3\CBR\8B\12\D3S\C1*\0A\8A\F2\D3\F5\8A\8A\01\9D\A9\1A\DA" }, section "__llvm_covmap", align 8
                                                                                                                                                                                                                                                                                                                                                      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/instrument-coverage/testprog.LINUX/testprog.ll:78:106: note: possible intended match here
@__profc__RNvCs2vQph7PjJMo_8testprog14will_be_called = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/instrument-coverage/testprog.LINUX/testprog.ll
Check file: /checkout/tests/codegen/instrument-coverage/testprog.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'testprog.1d4720e3d224b3e6-cgu.0' 
            2: source_filename = "testprog.1d4720e3d224b3e6-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            4: target triple = "x86_64-unknown-linux-gnu" 
            5:  
            6: %"core::fmt::rt::Argument<'_>" = type { %"core::fmt::rt::ArgumentType<'_>" } 
            7: %"core::fmt::rt::ArgumentType<'_>" = type { ptr, [1 x i64] } 
            8:  
            9: $__covrec_1D9B895249AD0C75u = comdat any 
           10:  
           11: $__covrec_36725C27D3039CB5u = comdat any 
           12:  
           13: $__covrec_1CE7CA97E6F1AF41u = comdat any 
           14:  
           15: $__covrec_A2CB80454FA0AB25u = comdat any 
           16:  
           17: $__covrec_9DA069E8A9CA2C9u = comdat any 
           18:  
           19: $__covrec_5690D14D8325DB4Au = comdat any 
           20:  
           21: $__covrec_138797266E6ABDC1u = comdat any 
           22:  
           23: $__covrec_53F9FD11B26D788Au = comdat any 
           24:  
           25: $__covrec_CEC584AB7289F43u = comdat any 
           26:  
           27: $__covrec_C6C88DA2C8EF10E1u = comdat any 
           28:  
           29: $__covrec_AAF7C95823D55CB4u = comdat any 
           30:  
           31: $__profc__RNvCs2vQph7PjJMo_8testprog14will_be_called = comdat nodeduplicate 
           32:  
           33: $__profc__RNvCs2vQph7PjJMo_8testprog18will_not_be_called = comdat nodeduplicate 
           34:  
           35: $__profc__RINvCs2vQph7PjJMo_8testprog5printRReEB2_ = comdat nodeduplicate 
           36:  
           37: $__profc__RINvCs2vQph7PjJMo_8testprog5printRbEB2_ = comdat nodeduplicate 
           38:  
           39: $__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4main0ReEB2_ = comdat nodeduplicate 
           40:  
           41: $__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains0_0bEB2_ = comdat nodeduplicate 
           42:  
           43: $__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains_0ReEB2_ = comdat nodeduplicate 
           44:  
           45: $__profc__RNvCs2vQph7PjJMo_8testprog4main = comdat nodeduplicate 
           46:  
           47: $__profc__RNCNvCs2vQph7PjJMo_8testprog4main0B3_ = comdat nodeduplicate 
           48:  
           49: $__profc__RNCNvCs2vQph7PjJMo_8testprog4mains_0B3_ = comdat nodeduplicate 
           50:  
           51: $__profc__RNCNvCs2vQph7PjJMo_8testprog4mains0_0B3_ = comdat nodeduplicate 
           52:  
           53: $__llvm_profile_filename = comdat any 
           54:  
           55: @0 = private unnamed_addr constant <{ [8 x i8], [8 x i8] }> <{ [8 x i8] zeroinitializer, [8 x i8] undef }>, align 8 
           56: @vtable.0 = private unnamed_addr constant <{ [24 x i8], ptr, ptr, ptr }> <{ [24 x i8] c"\00\00\00\00\00\00\00\00\08\00\00\00\00\00\00\00\08\00\00\00\00\00\00\00", ptr @_RNSNvYNCINvNtCslC1oO9aMz96_3std2rt10lang_startuE0INtNtNtCs4VswOIFEPhS_4core3ops8function6FnOnceuE9call_once6vtableCs2vQph7PjJMo_8testprog, ptr @_RNCINvNtCslC1oO9aMz96_3std2rt10lang_startuE0Cs2vQph7PjJMo_8testprog, ptr @_RNCINvNtCslC1oO9aMz96_3std2rt10lang_startuE0Cs2vQph7PjJMo_8testprog }>, align 8 
           57: @alloc_ca73ddfe2605aeea60911726d8b3dd82 = private unnamed_addr constant <{ [6 x i8] }> <{ [6 x i8] c"called" }>, align 1 
           58: @alloc_49a1e817e911805af64bbc7efb390101 = private unnamed_addr constant <{ [1 x i8] }> <{ [1 x i8] c"\0A" }>, align 1 
           59: @alloc_9771be2481f51be410bd2ac520d18601 = private unnamed_addr constant <{ ptr, [8 x i8], ptr, [8 x i8] }> <{ ptr inttoptr (i64 1 to ptr), [8 x i8] zeroinitializer, ptr @alloc_49a1e817e911805af64bbc7efb390101, [8 x i8] c"\01\00\00\00\00\00\00\00" }>, align 8 
           60: @alloc_6279f72697bc50b998c49d75e5a056ba = private unnamed_addr constant <{ [28 x i8] }> <{ [28 x i8] c"should not have been called\0A" }>, align 1 
           61: @alloc_be5c9986eadfae1b66bd7958e25dc482 = private unnamed_addr constant <{ ptr, [8 x i8] }> <{ ptr @alloc_6279f72697bc50b998c49d75e5a056ba, [8 x i8] c"\1C\00\00\00\00\00\00\00" }>, align 8 
           62: @alloc_8dfd08ba449915aba0baa3a82e12553a = private unnamed_addr constant <{ ptr, [8 x i8], ptr, [8 x i8], ptr, [8 x i8], ptr, [8 x i8] }> <{ ptr inttoptr (i64 1 to ptr), [8 x i8] zeroinitializer, ptr inttoptr (i64 1 to ptr), [8 x i8] zeroinitializer, ptr inttoptr (i64 1 to ptr), [8 x i8] zeroinitializer, ptr @alloc_49a1e817e911805af64bbc7efb390101, [8 x i8] c"\01\00\00\00\00\00\00\00" }>, align 8 
           63: @alloc_0fd49fb9c87d0ea55f07188f9eff834d = private unnamed_addr constant <{ [4 x i8] }> <{ [4 x i8] c" ***" }>, align 1 
           64: @alloc_9db824fd26015da05d288b42181536c6 = private unnamed_addr constant <{ [4 x i8] }> <{ [4 x i8] c"*** " }>, align 1 
           65: @alloc_3edf33e52be22c945259abc2c0759d11 = private unnamed_addr constant <{ [19 x i8] }> <{ [19 x i8] c"wrapped result is: " }>, align 1 
           66: @__covrec_1D9B895249AD0C75u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 2133449835006463093, i32 9, i64 -8922854311411313585, i64 -8916703881926681984, [9 x i8] c"\01\01\00\01\01 \01\04\02" }>, section "__llvm_covfun", comdat, align 8 
           67: @__covrec_36725C27D3039CB5u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 3923299551506832565, i32 9, i64 -5675644537888916618, i64 -8916703881926681984, [9 x i8] c"\01\01\00\01\01&\01\03\02" }>, section "__llvm_covfun", comdat, align 8 
           68: @__covrec_1CE7CA97E6F1AF41u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 2082856106445614913, i32 9, i64 -278832874628119237, i64 -8916703881926681984, [9 x i8] c"\01\01\00\01\01+\01\05\02" }>, section "__llvm_covfun", comdat, align 8 
           69: @__covrec_A2CB80454FA0AB25u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 -6716133384115934427, i32 9, i64 -278832874628119237, i64 -8916703881926681984, [9 x i8] c"\01\01\00\01\01+\01\05\02" }>, section "__llvm_covfun", comdat, align 8 
           70: @__covrec_9DA069E8A9CA2C9u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [26 x i8] }> <{ i64 709887169264394953, i32 26, i64 8001046714535999544, i64 -8916703881926681984, [26 x i8] c"\01\01\01\01\05\04\012\01\04\13\05\05\09\00\18\02\01\05\00\06\01\01\01\00\02" }>, section "__llvm_covfun", comdat, align 8 
           71: @__covrec_5690D14D8325DB4Au = linkonce_odr hidden constant <{ i64, i32, i64, i64, [26 x i8] }> <{ i64 6237715614751120202, i32 26, i64 8001046714535999544, i64 -8916703881926681984, [26 x i8] c"\01\01\01\01\05\04\012\01\04\13\05\05\09\00\18\02\01\05\00\06\01\01\01\00\02" }>, section "__llvm_covfun", comdat, align 8 
           72: @__covrec_138797266E6ABDC1u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [26 x i8] }> <{ i64 1407259599893609921, i32 26, i64 8001046714535999544, i64 -8916703881926681984, [26 x i8] c"\01\01\01\01\05\04\012\01\04\13\05\05\09\00\18\02\01\05\00\06\01\01\01\00\02" }>, section "__llvm_covfun", comdat, align 8 
           73: @__covrec_53F9FD11B26D788Au = linkonce_odr hidden constant <{ i64, i32, i64, i64, [41 x i8] }> <{ i64 6051145826798762122, i32 41, i64 -8278694896101811857, i64 -8916703881926681984, [41 x i8] c"\01\01\01\01\05\07\01;\01\04\13\05\04\14\012\05\01V\012\05\01V\01\06\02\01\0C\01/\02\01^\01\06\01\02\01\00\02" }>, section "__llvm_covfun", comdat, align 8 
           74: @__covrec_CEC584AB7289F43u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 931216300882566979, i32 9, i64 -8278694896101811857, i64 -8916703881926681984, [9 x i8] c"\01\01\00\01\01@:\00V" }>, section "__llvm_covfun", comdat, align 8 
           75: @__covrec_C6C88DA2C8EF10E1u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 -4122889728562360095, i32 9, i64 -8278694896101811857, i64 -8916703881926681984, [9 x i8] c"\01\01\00\01\01A:\00V" }>, section "__llvm_covfun", comdat, align 8 
           76: @__covrec_AAF7C95823D55CB4u = linkonce_odr hidden constant <{ i64, i32, i64, i64, [9 x i8] }> <{ i64 -6127207387618779980, i32 9, i64 -8278694896101811857, i64 -8916703881926681984, [9 x i8] c"\01\01\00\01\01C7\00^" }>, section "__llvm_covfun", comdat, align 8 
           77: @__llvm_coverage_mapping = private constant { { i32, i32, i32, i32 }, [69 x i8] } { { i32, i32, i32, i32 } { i32 0, i32 69, i32 0, i32 6 }, [69 x i8] c"\02FBx\DA\E3\D5O\CEHM\CE\CE/-\D1\CFO\CA2G\F0JR\8BK\8A\F5\93\F3SR\D3S\F3\F43\F3\8AK\8AJsS\F3Jt\93\F3\CBR\8B\12\D3S\C1*\0A\8A\F2\D3\F5\8A\8A\01\9D\A9\1A\DA" }, section "__llvm_covmap", align 8 
check:79'0                                                                                                                                                                                                                                                                                                                                                           X error: no match found
           78: @__profc__RNvCs2vQph7PjJMo_8testprog14will_be_called = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'1                                                                                                              ?                                   possible intended match
           79: @__profd__RNvCs2vQph7PjJMo_8testprog14will_be_called = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 2133449835006463093, i64 -8922854311411313585, i64 sub (i64 ptrtoint (ptr @__profc__RNvCs2vQph7PjJMo_8testprog14will_be_called to i64), i64 ptrtoint (ptr @__profd__RNvCs2vQph7PjJMo_8testprog14will_be_called to i64)), i64 0, ptr null, ptr null, i32 1, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RNvCs2vQph7PjJMo_8testprog14will_be_called), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           80: @__profc__RNvCs2vQph7PjJMo_8testprog18will_not_be_called = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           81: @__profd__RNvCs2vQph7PjJMo_8testprog18will_not_be_called = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 3923299551506832565, i64 -5675644537888916618, i64 sub (i64 ptrtoint (ptr @__profc__RNvCs2vQph7PjJMo_8testprog18will_not_be_called to i64), i64 ptrtoint (ptr @__profd__RNvCs2vQph7PjJMo_8testprog18will_not_be_called to i64)), i64 0, ptr null, ptr null, i32 1, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RNvCs2vQph7PjJMo_8testprog18will_not_be_called), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           82: @__profc__RINvCs2vQph7PjJMo_8testprog5printRReEB2_ = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           83: @__profd__RINvCs2vQph7PjJMo_8testprog5printRReEB2_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 2082856106445614913, i64 -278832874628119237, i64 sub (i64 ptrtoint (ptr @__profc__RINvCs2vQph7PjJMo_8testprog5printRReEB2_ to i64), i64 ptrtoint (ptr @__profd__RINvCs2vQph7PjJMo_8testprog5printRReEB2_ to i64)), i64 0, ptr null, ptr null, i32 1, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RINvCs2vQph7PjJMo_8testprog5printRReEB2_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           84: @__profc__RINvCs2vQph7PjJMo_8testprog5printRbEB2_ = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           85: @__profd__RINvCs2vQph7PjJMo_8testprog5printRbEB2_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 -6716133384115934427, i64 -278832874628119237, i64 sub (i64 ptrtoint (ptr @__profc__RINvCs2vQph7PjJMo_8testprog5printRbEB2_ to i64), i64 ptrtoint (ptr @__profd__RINvCs2vQph7PjJMo_8testprog5printRbEB2_ to i64)), i64 0, ptr null, ptr null, i32 1, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RINvCs2vQph7PjJMo_8testprog5printRbEB2_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           86: @__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4main0ReEB2_ = private global [2 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           87: @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4main0ReEB2_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 709887169264394953, i64 8001046714535999544, i64 sub (i64 ptrtoint (ptr @__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4main0ReEB2_ to i64), i64 ptrtoint (ptr @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4main0ReEB2_ to i64)), i64 0, ptr null, ptr null, i32 2, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4main0ReEB2_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           88: @__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains0_0bEB2_ = private global [2 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           89: @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains0_0bEB2_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 6237715614751120202, i64 8001046714535999544, i64 sub (i64 ptrtoint (ptr @__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains0_0bEB2_ to i64), i64 ptrtoint (ptr @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains0_0bEB2_ to i64)), i64 0, ptr null, ptr null, i32 2, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains0_0bEB2_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           90: @__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains_0ReEB2_ = private global [2 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           91: @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains_0ReEB2_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 1407259599893609921, i64 8001046714535999544, i64 sub (i64 ptrtoint (ptr @__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains_0ReEB2_ to i64), i64 ptrtoint (ptr @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains_0ReEB2_ to i64)), i64 0, ptr null, ptr null, i32 2, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains_0ReEB2_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           92: @__profc__RNvCs2vQph7PjJMo_8testprog4main = private global [2 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           93: @__profd__RNvCs2vQph7PjJMo_8testprog4main = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 6051145826798762122, i64 -8278694896101811857, i64 sub (i64 ptrtoint (ptr @__profc__RNvCs2vQph7PjJMo_8testprog4main to i64), i64 ptrtoint (ptr @__profd__RNvCs2vQph7PjJMo_8testprog4main to i64)), i64 0, ptr null, ptr null, i32 2, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RNvCs2vQph7PjJMo_8testprog4main), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           94: @__profc__RNCNvCs2vQph7PjJMo_8testprog4main0B3_ = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           95: @__profd__RNCNvCs2vQph7PjJMo_8testprog4main0B3_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 931216300882566979, i64 -8278694896101811857, i64 sub (i64 ptrtoint (ptr @__profc__RNCNvCs2vQph7PjJMo_8testprog4main0B3_ to i64), i64 ptrtoint (ptr @__profd__RNCNvCs2vQph7PjJMo_8testprog4main0B3_ to i64)), i64 0, ptr null, ptr null, i32 1, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RNCNvCs2vQph7PjJMo_8testprog4main0B3_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           96: @__profc__RNCNvCs2vQph7PjJMo_8testprog4mains_0B3_ = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           97: @__profd__RNCNvCs2vQph7PjJMo_8testprog4mains_0B3_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 -4122889728562360095, i64 -8278694896101811857, i64 sub (i64 ptrtoint (ptr @__profc__RNCNvCs2vQph7PjJMo_8testprog4mains_0B3_ to i64), i64 ptrtoint (ptr @__profd__RNCNvCs2vQph7PjJMo_8testprog4mains_0B3_ to i64)), i64 0, ptr null, ptr null, i32 1, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RNCNvCs2vQph7PjJMo_8testprog4mains_0B3_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           98: @__profc__RNCNvCs2vQph7PjJMo_8testprog4mains0_0B3_ = private global [1 x i64] zeroinitializer, section "__llvm_prf_cnts", comdat, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           99: @__profd__RNCNvCs2vQph7PjJMo_8testprog4mains0_0B3_ = private global { i64, i64, i64, i64, ptr, ptr, i32, [2 x i16], i32 } { i64 -6127207387618779980, i64 -8278694896101811857, i64 sub (i64 ptrtoint (ptr @__profc__RNCNvCs2vQph7PjJMo_8testprog4mains0_0B3_ to i64), i64 ptrtoint (ptr @__profd__RNCNvCs2vQph7PjJMo_8testprog4mains0_0B3_ to i64)), i64 0, ptr null, ptr null, i32 1, [2 x i16] zeroinitializer, i32 0 }, section "__llvm_prf_data", comdat($__profc__RNCNvCs2vQph7PjJMo_8testprog4mains0_0B3_), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          100: @__llvm_prf_nm = private constant [145 x i8] c"\FB\03\8D\01x\DA\9D\CD\D1\0A\C2 \14\80az\A3\D3Z\B4n']\14$\CD\178\B8\92f\98\8A\8A\BE\FE\967\11\B4%\DD\7F??2\1A\89\AFbg\87\DD\E5q:\1Bl\82\F0\C1:s_\D7I*\85\BD\C0+WJ\DCV\B8`\9Bl\B5\09\1F\FE8\1Bl\AD\93:0&\0Em\85%\B2\FF\09\F7\C9q\8BI\86\81\12\1A']?\B9\D4P\B0\F8^z@\F8\F7\EA\F1\FD\9DM\B3|\01\B2L\A0\DD`\01\9B\9E\85\10\B2\1C\01\DFd\B3\A1", section "__llvm_prf_names", align 1 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          101: @llvm.compiler.used = appending global [11 x ptr] [ptr @__profd__RNvCs2vQph7PjJMo_8testprog14will_be_called, ptr @__profd__RNvCs2vQph7PjJMo_8testprog18will_not_be_called, ptr @__profd__RINvCs2vQph7PjJMo_8testprog5printRReEB2_, ptr @__profd__RINvCs2vQph7PjJMo_8testprog5printRbEB2_, ptr @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4main0ReEB2_, ptr @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains0_0bEB2_, ptr @__profd__RINvCs2vQph7PjJMo_8testprog9wrap_withNCNvB2_4mains_0ReEB2_, ptr @__profd__RNvCs2vQph7PjJMo_8testprog4main, ptr @__profd__RNCNvCs2vQph7PjJMo_8testprog4main0B3_, ptr @__profd__RNCNvCs2vQph7PjJMo_8testprog4mains_0B3_, ptr @__profd__RNCNvCs2vQph7PjJMo_8testprog4mains0_0B3_], section "llvm.metadata" 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          102: @llvm.used = appending global [13 x ptr] [ptr @__covrec_1D9B895249AD0C75u, ptr @__covrec_36725C27D3039CB5u, ptr @__covrec_1CE7CA97E6F1AF41u, ptr @__covrec_A2CB80454FA0AB25u, ptr @__covrec_9DA069E8A9CA2C9u, ptr @__covrec_5690D14D8325DB4Au, ptr @__covrec_138797266E6ABDC1u, ptr @__covrec_53F9FD11B26D788Au, ptr @__covrec_CEC584AB7289F43u, ptr @__covrec_C6C88DA2C8EF10E1u, ptr @__covrec_AAF7C95823D55CB4u, ptr @__llvm_coverage_mapping, ptr @__llvm_prf_nm], section "llvm.metadata" 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          103: @__llvm_profile_filename = hidden constant [22 x i8] c"default_%m_%p.profraw\00", comdat 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'0     ~
check:79'0     ~
          105: ; <core::fmt::Arguments>::new_v1::<2, 1> 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          106: ; Function Attrs: inlinehint nonlazybind uwtable 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          107: define internal void @_RINvMs2_NtCs4VswOIFEPhS_4core3fmtNtB6_9Arguments6new_v1Kj2_Kj1_ECs2vQph7PjJMo_8testprog(ptr sret([48 x i8]) align 8 %_0, ptr align 8 %pieces, ptr align 8 %args) unnamed_addr #0 { 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'0     ~~~~~~~
          109:  %0 = getelementptr inbounds i8, ptr %_0, i64 16 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          110:  store ptr %pieces, ptr %0, align 8 
          110:  store ptr %pieces, ptr %0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          111:  %1 = getelementptr inbounds i8, ptr %0, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          112:  store i64 2, ptr %1, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          113:  %2 = load ptr, ptr @0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          114:  %3 = load i64, ptr getelementptr inbounds (i8, ptr @0, i64 8), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          115:  store ptr %2, ptr %_0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          116:  %4 = getelementptr inbounds i8, ptr %_0, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          117:  store i64 %3, ptr %4, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          118:  %5 = getelementptr inbounds i8, ptr %_0, i64 32 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          119:  store ptr %args, ptr %5, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          120:  %6 = getelementptr inbounds i8, ptr %5, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          121:  store i64 1, ptr %6, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          122:  ret void 
check:79'0     ~~~~~~~~~~
          123: } 
check:79'0     ~~
check:79'0     ~
check:79'0     ~
          125: ; <core::fmt::Arguments>::new_v1::<4, 3> 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          126: ; Function Attrs: inlinehint nonlazybind uwtable 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          127: define internal void @_RINvMs2_NtCs4VswOIFEPhS_4core3fmtNtB6_9Arguments6new_v1Kj4_Kj3_ECs2vQph7PjJMo_8testprog(ptr sret([48 x i8]) align 8 %_0, ptr align 8 %pieces, ptr align 8 %args) unnamed_addr #0 { 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'0     ~~~~~~~
          129:  %0 = getelementptr inbounds i8, ptr %_0, i64 16 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          130:  store ptr %pieces, ptr %0, align 8 
          130:  store ptr %pieces, ptr %0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          131:  %1 = getelementptr inbounds i8, ptr %0, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          132:  store i64 4, ptr %1, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          133:  %2 = load ptr, ptr @0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          134:  %3 = load i64, ptr getelementptr inbounds (i8, ptr @0, i64 8), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          135:  store ptr %2, ptr %_0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          136:  %4 = getelementptr inbounds i8, ptr %_0, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          137:  store i64 %3, ptr %4, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          138:  %5 = getelementptr inbounds i8, ptr %_0, i64 32 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          139:  store ptr %args, ptr %5, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          140:  %6 = getelementptr inbounds i8, ptr %5, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          141:  store i64 3, ptr %6, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          142:  ret void 
check:79'0     ~~~~~~~~~~
          143: } 
check:79'0     ~~
check:79'0     ~
check:79'0     ~
          145: ; <core::fmt::Arguments>::new_const::<1> 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          146: ; Function Attrs: inlinehint nonlazybind uwtable 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          147: define internal void @_RINvMs2_NtCs4VswOIFEPhS_4core3fmtNtB6_9Arguments9new_constKj1_ECs2vQph7PjJMo_8testprog(ptr sret([48 x i8]) align 8 %_0, ptr align 8 %pieces) unnamed_addr #0 { 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'0     ~~~~~~~
          149:  %0 = getelementptr inbounds i8, ptr %_0, i64 16 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          150:  store ptr %pieces, ptr %0, align 8 
          150:  store ptr %pieces, ptr %0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          151:  %1 = getelementptr inbounds i8, ptr %0, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          152:  store i64 1, ptr %1, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          153:  %2 = load ptr, ptr @0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          154:  %3 = load i64, ptr getelementptr inbounds (i8, ptr @0, i64 8), align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          155:  store ptr %2, ptr %_0, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          156:  %4 = getelementptr inbounds i8, ptr %_0, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          157:  store i64 %3, ptr %4, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          158:  %5 = getelementptr inbounds i8, ptr %_0, i64 32 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          159:  store ptr inttoptr (i64 8 to ptr), ptr %5, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          160:  %6 = getelementptr inbounds i8, ptr %5, i64 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          161:  store i64 0, ptr %6, align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          162:  ret void 
check:79'0     ~~~~~~~~~~
          163: } 
check:79'0     ~~
check:79'0     ~
          165: ; core::ptr::drop_in_place::<std::rt::lang_start<()>::{closure#0}> 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          166: ; Function Attrs: inlinehint nonlazybind uwtable 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          167: define internal void @_RINvNtCs4VswOIFEPhS_4core3ptr13drop_in_placeNCINvNtCslC1oO9aMz96_3std2rt10lang_startuE0ECs2vQph7PjJMo_8testprog(ptr align 8 %_1) unnamed_addr #0 { 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'0     ~~~~~~~
          169:  ret void 
check:79'0     ~~~~~~~~~~
          170: } 
          170: } 
check:79'0     ~~
          171:  
check:79'0     ~
          172: ; std::rt::lang_start::<()> 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          173: ; Function Attrs: nonlazybind uwtable 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          174: define hidden i64 @_RINvNtCslC1oO9aMz96_3std2rt10lang_startuECs2vQph7PjJMo_8testprog(ptr %main, i64 %argc, ptr %argv, i8 %sigpipe) unnamed_addr #1 { 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:79'0     ~~~~~~~
          176:  %_8 = alloca [8 x i8], align 8 
check:79'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          177:  %_5 = alloca [8 x i8], align 8 
---


Some tests failed in compiletest suite=
failures:
codegen     mode=[codegen] tests/codegen/instrument-coverage/testprog.rs#LINUXcodegen
 host=
test result: x86_64-unknown-linux-gnu target=FAILEDx86_64-unknown-linux-gnu

Build completed unsuccessfully in 0:18:00
  local time: Wed Dec 11 07:56:00 UTC 2024
  network time: Wed, 11 Dec 2024 07:56:00 GMT
```

</details>

--- a/compiler/rustc_codegen_llvm/src/coverageinfo/ffi.rs
+++ b/compiler/rustc_codegen_llvm/src/coverageinfo/ffi.rs
@@ -152,6 +152,34 @@ impl CoverageSpan {
     }
 }
 
+/// Holds tables of the various region types in one struct.
+///
+/// Don't pass this struct across FFI; pass the individual region tables as
+/// pointer/length pairs instead.
+///
+/// Each field name has a `_regions` suffix for improved readability after
+/// exhaustive destructing, which ensures that all region types are handled.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct Regions {
+    pub(crate) code_regions: Vec<CodeRegion>,
+    pub(crate) branch_regions: Vec<BranchRegion>,
+    pub(crate) mcdc_branch_regions: Vec<MCDCBranchRegion>,
+    pub(crate) mcdc_decision_regions: Vec<MCDCDecisionRegion>,
+}
+
+impl Regions {
+    /// Returns true if none of this structure's tables contain any regions.
+    pub(crate) fn has_no_regions(&self) -> bool {
+        let Self { code_regions, branch_regions, mcdc_branch_regions, mcdc_decision_regions } =
+            self;
+
+        code_regions.is_empty()
+            && branch_regions.is_empty()
+            && mcdc_branch_regions.is_empty()
+            && mcdc_decision_regions.is_empty()
+    }
+}
+
 /// Must match the layout of `LLVMRustCoverageCodeRegion`.
 #[derive(Clone, Debug)]
 #[repr(C)]

--- a/compiler/rustc_codegen_llvm/src/coverageinfo/llvm_cov.rs
+++ b/compiler/rustc_codegen_llvm/src/coverageinfo/llvm_cov.rs
@@ -62,11 +62,10 @@ pub(crate) fn write_filenames_to_buffer<'a>(
 pub(crate) fn write_function_mappings_to_buffer(
     virtual_file_mapping: &[u32],
     expressions: &[ffi::CounterExpression],
-    code_regions: &[ffi::CodeRegion],
-    branch_regions: &[ffi::BranchRegion],
-    mcdc_branch_regions: &[ffi::MCDCBranchRegion],
-    mcdc_decision_regions: &[ffi::MCDCDecisionRegion],
+    regions: &ffi::Regions,
 ) -> Vec<u8> {
+    let ffi::Regions { code_regions, branch_regions, mcdc_branch_regions, mcdc_decision_regions } =
+        regions;
     llvm::build_byte_buffer(|buffer| unsafe {
         llvm::LLVMRustCoverageWriteFunctionMappingsToBuffer(
             virtual_file_mapping.as_ptr(),

--- a/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs
+++ b/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs
@@ -194,7 +194,7 @@ rustc_index::newtype_index! {
 
 /// Holds a mapping from "local" (per-function) file IDs to "global" (per-CGU)
 /// file IDs.
-#[derive(Default)]
+#[derive(Debug, Default)]
 struct VirtualFileMapping {
     local_to_global: IndexVec<LocalFileId, GlobalFileId>,
     global_to_local: FxIndexMap<GlobalFileId, LocalFileId>,
@@ -208,10 +208,10 @@ impl VirtualFileMapping {
             .or_insert_with(|| self.local_to_global.push(global_file_id))
     }
 
-    fn into_vec(self) -> Vec<u32> {
-        // This conversion should be optimized away to ~zero overhead.
-        // In any case, it's probably not hot enough to worry about.
-        self.local_to_global.into_iter().map(|global| global.as_u32()).collect()
+    fn to_vec(&self) -> Vec<u32> {
+        // This clone could be avoided by transmuting `&[GlobalFileId]` to `&[u32]`,
+        // but it isn't hot or expensive enough to justify the extra unsafety.
+        self.local_to_global.iter().map(|&global| GlobalFileId::as_u32(global)).collect()
     }
 }
 

--- a/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs
+++ b/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs
@@ -80,10 +80,6 @@ pub(crate) fn finalize(cx: &CodegenCx<'_, '_>) {
     let filenames_val = cx.const_bytes(&filenames_buffer);
     let filenames_ref = llvm_cov::hash_bytes(&filenames_buffer);
 
-    // Generate the coverage map header, which contains the filenames used by
-    // this CGU's coverage mappings, and store it in a well-known global.
-    generate_covmap_record(cx, covmap_version, filenames_size, filenames_val);
-
     let mut unused_function_names = Vec::new();
 
     let covfun_records = function_coverage_map
@@ -92,6 +88,15 @@ pub(crate) fn finalize(cx: &CodegenCx<'_, '_>) {
             prepare_covfun_record(tcx, &global_file_table, instance, &function_coverage)
         })
         .collect::<Vec<_>>();
+
+    // If there are no covfun records for this CGU, don't generate a covmap record.
+    // Emitting a covmap record without any covfun records causes `llvm-cov` to
+    // fail when generating coverage reports, and if there are no covfun records
+    // then the covmap record isn't useful anyway.
+    // This should prevent a repeat of <https://github.com/rust-lang/rust/issues/133606>.
+    if covfun_records.is_empty() {
+        return;
+    }
 
     for covfun in &covfun_records {
         unused_function_names.extend(covfun.mangled_function_name_if_unused());
@@ -117,6 +122,11 @@ pub(crate) fn finalize(cx: &CodegenCx<'_, '_>) {
         llvm::set_linkage(array, llvm::Linkage::InternalLinkage);
         llvm::set_initializer(array, initializer);
     }
+
+    // Generate the coverage map header, which contains the filenames used by
+    // this CGU's coverage mappings, and store it in a well-known global.
+    // (This is skipped if we returned early due to having no covfun records.)
+    generate_covmap_record(cx, covmap_version, filenames_size, filenames_val);
 }
 
 /// Maps "global" (per-CGU) file ID numbers to their underlying filenames.

--- a/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs
+++ b/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs
@@ -29,7 +29,10 @@ pub(crate) struct CovfunRecord<'tcx> {
     mangled_function_name: &'tcx str,
     source_hash: u64,
     is_used: bool,
-    coverage_mapping_buffer: Vec<u8>,
+
+    virtual_file_mapping: VirtualFileMapping,
+    expressions: Vec<ffi::CounterExpression>,
+    regions: ffi::Regions,
 }
 
 impl<'tcx> CovfunRecord<'tcx> {
@@ -46,50 +49,40 @@ pub(crate) fn prepare_covfun_record<'tcx>(
     instance: Instance<'tcx>,
     function_coverage: &FunctionCoverage<'tcx>,
 ) -> Option<CovfunRecord<'tcx>> {
-    let mangled_function_name = tcx.symbol_name(instance).name;
-    let source_hash = function_coverage.source_hash();
-    let is_used = function_coverage.is_used();
+    let mut covfun = CovfunRecord {
+        mangled_function_name: tcx.symbol_name(instance).name,
+        source_hash: function_coverage.source_hash(),
+        is_used: function_coverage.is_used(),
+        virtual_file_mapping: VirtualFileMapping::default(),
+        expressions: function_coverage.counter_expressions().collect::<Vec<_>>(),
+        regions: ffi::Regions::default(),
+    };
 
-    let coverage_mapping_buffer =
-        encode_mappings_for_function(tcx, global_file_table, function_coverage);
+    fill_region_tables(tcx, global_file_table, function_coverage, &mut covfun);
 
-    if coverage_mapping_buffer.is_empty() {
-        if function_coverage.is_used() {
-            bug!(
-                "A used function should have had coverage mapping data but did not: {}",
-                mangled_function_name
-            );
+    if covfun.regions.has_no_regions() {
+        if covfun.is_used {
+            bug!("a used function should have had coverage mapping data but did not: {covfun:?}");
         } else {
-            debug!("unused function had no coverage mapping data: {}", mangled_function_name);
+            debug!(?covfun, "unused function had no coverage mapping data");
             return None;
         }
     }
 
-    Some(CovfunRecord { mangled_function_name, source_hash, is_used, coverage_mapping_buffer })
+    Some(covfun)
 }
 
-/// Using the expressions and counter regions collected for a single function,
-/// generate the variable-sized payload of its corresponding `__llvm_covfun`
-/// entry. The payload is returned as a vector of bytes.
-///
-/// Newly-encountered filenames will be added to the global file table.
-fn encode_mappings_for_function(
-    tcx: TyCtxt<'_>,
+/// Populates the mapping region tables in the current function's covfun record.
+fn fill_region_tables<'tcx>(
+    tcx: TyCtxt<'tcx>,
     global_file_table: &GlobalFileTable,
-    function_coverage: &FunctionCoverage<'_>,
-) -> Vec<u8> {
+    function_coverage: &FunctionCoverage<'tcx>,
+    covfun: &mut CovfunRecord<'tcx>,
+) {
     let counter_regions = function_coverage.counter_regions();
     if counter_regions.is_empty() {
-        return Vec::new();
+        return;
     }
-
-    let expressions = function_coverage.counter_expressions().collect::<Vec<_>>();
-
-    let mut virtual_file_mapping = VirtualFileMapping::default();
-    let mut code_regions = vec![];
-    let mut branch_regions = vec![];
-    let mut mcdc_branch_regions = vec![];
-    let mut mcdc_decision_regions = vec![];
 
     // Currently a function's mappings must all be in the same file as its body span.
     let file_name = span_file_name(tcx, function_coverage.function_coverage_info.body_span);
@@ -98,8 +91,11 @@ fn encode_mappings_for_function(
     let global_file_id = global_file_table.global_file_id_for_file_name(file_name);
 
     // Associate that global file ID with a local file ID for this function.
-    let local_file_id = virtual_file_mapping.local_id_for_global(global_file_id);
+    let local_file_id = covfun.virtual_file_mapping.local_id_for_global(global_file_id);
     debug!("  file id: {local_file_id:?} => {global_file_id:?} = '{file_name:?}'");
+
+    let ffi::Regions { code_regions, branch_regions, mcdc_branch_regions, mcdc_decision_regions } =
+        &mut covfun.regions;
 
     // For each counter/region pair in this function+file, convert it to a
     // form suitable for FFI.
@@ -133,16 +129,6 @@ fn encode_mappings_for_function(
             }
         }
     }
-
-    // Encode the function's coverage mappings into a buffer.
-    llvm_cov::write_function_mappings_to_buffer(
-        &virtual_file_mapping.into_vec(),
-        &expressions,
-        &code_regions,
-        &branch_regions,
-        &mcdc_branch_regions,
-        &mcdc_decision_regions,
-    )
 }
 
 /// Generates the contents of the covfun record for this function, which
@@ -157,8 +143,17 @@ pub(crate) fn generate_covfun_record<'tcx>(
         mangled_function_name,
         source_hash,
         is_used,
-        ref coverage_mapping_buffer, // Previously-encoded coverage mappings
+        ref virtual_file_mapping,
+        ref expressions,
+        ref regions,
     } = covfun;
+
+    // Encode the function's coverage mappings into a buffer.
+    let coverage_mapping_buffer = llvm_cov::write_function_mappings_to_buffer(
+        &virtual_file_mapping.to_vec(),
+        expressions,
+        regions,
+    );
 
     // Concatenate the encoded coverage mappings
     let coverage_mapping_size = coverage_mapping_buffer.len();

--- a/tests/codegen/instrument-coverage/testprog.rs
+++ b/tests/codegen/instrument-coverage/testprog.rs
@@ -73,11 +73,9 @@ fn main() {
 
 // WIN:          $__llvm_profile_runtime_user = comdat any
 
-// CHECK:        @__llvm_coverage_mapping = private constant
-// CHECK-SAME:   section "[[INSTR_PROF_COVMAP]]", align 8
+// CHECK-DAG:    @__llvm_coverage_mapping = private constant {{.*}}, section "[[INSTR_PROF_COVMAP]]", align 8
 
-// CHECK:        @__covrec_{{[A-F0-9]+}}u = linkonce_odr hidden constant
-// CHECK-SAME:   section "[[INSTR_PROF_COVFUN]]"[[COMDAT_IF_SUPPORTED]], align 8
+// CHECK-DAG:    @__covrec_{{[A-F0-9]+}}u = linkonce_odr hidden constant {{.*}}, section "[[INSTR_PROF_COVFUN]]"[[COMDAT_IF_SUPPORTED]], align 8
 
 // WIN:          @__llvm_profile_runtime = external{{.*}}global i32
 

